processor/src/advice/mem_provider.rs

```diff
@@ -113,53 +104,28 @@ impl AdviceProvider for MemAdviceProvider {
         depth: &Felt,
         index: &Felt,
     ) -> Result<MerklePath, ExecutionError> {
-        // look up the merkle set and return an error if none is found
-        let merkle_set = self
-            .sets
-            .get(&root.into_bytes())
-            .ok_or_else(|| ExecutionError::MerkleSetNotFound(root.into_bytes()))?;
-
-        // get the Merkle path from the merkle set based on depth and index
         let index = NodeIndex::from_elements(depth, index)
-            .map_err(ExecutionError::MerkleSetLookupFailed)?;
-        let path = merkle_set.get_path(index).map_err(ExecutionError::MerkleSetLookupFailed)?;
-
-        Ok(path)
+            .map_err(|_| ExecutionError::MerkleNodeIndex(*depth, *index))?;
+        self.store
+            .get_path(root, index)
+            .map(|(_leaf, path)| path)
+            .map_err(ExecutionError::MerkleSetLookupFailed)
     }
 
-    fn update_merkle_leaf(
+    fn update_merkle_node(
         &mut self,
         root: Word,
-        index: Felt,
-        leaf_value: Word,
-        update_in_copy: bool,
+        depth: &Felt,
+        index: &Felt,
+        value: Word,
     ) -> Result<MerklePath, ExecutionError> {
-        // look up the merkle set and return error if none is found. if we are updating a copy,
-        // clone the merkle set; otherwise remove it from the map because the root will change,
-        // and we'll re-insert the set later under a different root.
-        let mut merkle_set = if update_in_copy {
-            // look up the merkle set and return an error if none is found
-            self.sets
-                .get(&root.into_bytes())
-                .ok_or_else(|| ExecutionError::MerkleSetNotFound(root.into_bytes()))?
-                .clone()
-        } else {
-            self.sets
-                .remove(&root.into_bytes())
-                .ok_or_else(|| ExecutionError::MerkleSetNotFound(root.into_bytes()))?
-        };
-
-        // get the Merkle path from the merkle set for the leaf at the specified index
-        let index = NodeIndex::new(merkle_set.depth(), index.as_int());
-        let path = merkle_set.get_path(index).map_err(ExecutionError::MerkleSetLookupFailed)?;
-
-        // update the merkle set and re-insert it into the map
-        merkle_set
-            .update_leaf(index.value(), leaf_value)
+        let merkle_index = NodeIndex::from_elements(depth, index)
+            .map_err(|_| ExecutionError::MerkleNodeIndex(*depth, *index))?;
+        let new_root = self
+            .store
+            .set_node(root, merkle_index, value)
             .map_err(ExecutionError::MerkleSetLookupFailed)?;
-        self.sets.insert(merkle_set.root().into_bytes(), merkle_set);
-
-        Ok(path)
+        self.get_merkle_path(new_root, depth, index)
```

Comment:
This will result in traversing the tree twice as we do `get_merkle_path()` as a part of `set_node()`. So, I'm thinking we should modify `set_node()` to return a tuple `(root, path)`.

Also, nit: I'd probably change `merkle_index` into just `node_index`.

Reply:
As this is not performed in this repo, added as `nice to have`

Reply:
I think it is fine not to do now, but we should do this before we merge into `next` and we should also make the corresponding change in `miden-crypto`.